controllers/release/release_adapter.go

```diff
@@ -228,24 +229,31 @@ func (a *Adapter) EnsureSnapshotEnvironmentBindingIsCreated() (results.Operation
 // EnsureTargetContextIsSet is an operation that will ensure that the targetContext of the adapter is not nil. It will
 // set it to a context from the workspace in the ReleasePlan if one is defined or to its own context if not.
 func (a *Adapter) EnsureTargetContextIsSet() (results.OperationResult, error) {
-	if a.targetContext == nil {
+	if _, isKcp := logicalcluster.ClusterFromContext(a.context); !isKcp {
+		a.targetContext = a.context
+		return results.ContinueProcessing()
+	}
+
+	// If the target namespace/workspace is not set we should get the target from the ReleasePlan
+	if a.release.Status.Target == (kcp.NamespaceReference{}) {
 		releasePlan, err := a.getReleasePlan()
 		if err != nil {
 			patch := client.MergeFrom(a.release.DeepCopy())
 			a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, err.Error())
 			return results.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.release, patch))
 		}
 
-		if releasePlan.Spec.Target.Workspace == "" {
-			a.logger.Info("Running in a traditional kubernetes environment")
-			a.targetContext = a.context
-		} else {
-			a.logger.Info("Running in a KCP environment")
-			a.targetContext = logicalcluster.WithCluster(a.context, logicalcluster.New(releasePlan.Spec.Target.Workspace))
+		patch := client.MergeFrom(a.release.DeepCopy())
+		a.release.Status.Target.Workspace = releasePlan.Spec.Target.Workspace
+		a.release.Status.Target.Namespace = releasePlan.Spec.Target.Namespace
+		err = a.client.Status().Patch(a.context, a.release, patch)
+		if err != nil {
+			return results.RequeueOnErrorOrStop(err)
 		}
-
-		a.syncer.SetContext(a.targetContext)
+		a.targetContext = logicalcluster.WithCluster(a.context, logicalcluster.New(a.release.Status.Target.Workspace))
```

Comment:
This like should be moved outside the block. If not, the `targetContext` will get set only for non-KCP and whenever the Release status doesn't have the target set.